### ISSUE TITLE
HttpClient NPE in SslClientConnectionFactory

### ISSUE THREAD

**Issue by sslavic** [eclipse/jetty.project#1109]
Couldn't reproduce issue with jetty 9.3.9.v20160517.
Reproduced issue with 9.3.10.v20160621 and with latest 9.3.14.v20161028. Didn't try intermediate releases.

Stack trace with 9.3.10:
```
12:47:56 [DEBUG] [org.eclipse.jetty.io.ManagedSelector] []
java.lang.NullPointerException: null
	at org.eclipse.jetty.io.ssl.SslClientConnectionFactory.customize(SslClientConnectionFactory.java:87) ~[jetty-io-9.3.10.v20160621.jar:9.3.10.v20160621]
	at org.eclipse.jetty.io.ssl.SslClientConnectionFactory.newConnection(SslClientConnectionFactory.java:66) ~[jetty-io-9.3.10.v20160621.jar:9.3.10.v20160621]
	at org.eclipse.jetty.client.AbstractHttpClientTransport$ClientSelectorManager.newConnection(AbstractHttpClientTransport.java:187) ~[jetty-client-9.3.6.v20151106.jar:9.3.6.v20151106]
	at org.eclipse.jetty.io.ManagedSelector.createEndPoint(ManagedSelector.java:422) [jetty-io-9.3.10.v20160621.jar:9.3.10.v20160621]
	at org.eclipse.jetty.io.ManagedSelector.access$1600(ManagedSelector.java:56) [jetty-io-9.3.10.v20160621.jar:9.3.10.v20160621]
	at org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run(ManagedSelector.java:605) ~[jetty-io-9.3.10.v20160621.jar:9.3.10.v20160621]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303) [jetty-util-9.3.10.v20160621.jar:9.3.10.v20160621]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148) [jetty-util-9.3.10.v20160621.jar:9.3.10.v20160621]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.execute(ExecuteProduceConsume.java:100) [jetty-util-9.3.10.v20160621.jar:9.3.10.v20160621]
	at org.eclipse.jetty.io.ManagedSelector.run(ManagedSelector.java:147) [jetty-io-9.3.10.v20160621.jar:9.3.10.v20160621]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_91]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_91]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_91]
```

Stack trace with 9.3.14:
```
10:17:13 [DEBUG] [org.eclipse.jetty.io.ManagedSelector] []
java.lang.NullPointerException: null
	at org.eclipse.jetty.io.ssl.SslClientConnectionFactory.customize(SslClientConnectionFactory.java:87) ~[jetty-io-9.3.14.v20161028.jar:9.3.14.v20161028]
	at org.eclipse.jetty.io.ssl.SslClientConnectionFactory.newConnection(SslClientConnectionFactory.java:66) ~[jetty-io-9.3.14.v20161028.jar:9.3.14.v20161028]
	at org.eclipse.jetty.client.AbstractHttpClientTransport$ClientSelectorManager.newConnection(AbstractHttpClientTransport.java:187) ~[jetty-client-9.3.6.v20151106.jar:9.3.6.v20151106]
	at org.eclipse.jetty.io.ManagedSelector.createEndPoint(ManagedSelector.java:422) ~[jetty-io-9.3.14.v20161028.jar:9.3.14.v20161028]
	at org.eclipse.jetty.io.ManagedSelector.access$1600(ManagedSelector.java:56) ~[jetty-io-9.3.14.v20161028.jar:9.3.14.v20161028]
	at org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run(ManagedSelector.java:605) ~[jetty-io-9.3.14.v20161028.jar:9.3.14.v20161028]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303) [jetty-util-9.3.14.v20161028.jar:9.3.14.v20161028]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148) [jetty-util-9.3.14.v20161028.jar:9.3.14.v20161028]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136) [jetty-util-9.3.14.v20161028.jar:9.3.14.v20161028]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_91]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_91]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_91]
```

**Comment by sbordet** [comment]
How do you reproduce this ?

**Comment by sslavic** [comment]
On particular test environment, HttpClient has SslContextFactory injected that has trustAll flag on true, making a call to server with self-signed certificate.

**Comment by sbordet** [comment]
The line that throws the NPE has a `null` connector pulled out from the context map. This should never happen, and to my knowledge does not happen in the Jetty test suite where we exactly use an all trusting `SslContextFactory` and self-signed certificates.

Can you post a simple test case that reproduces the issue ? I would like to see the code.

**Comment by sslavic** [comment]
Noticed we had jetty-server and jetty-client dependency, both in upgrade on different version, jetty-server always on 9.3.9.v20160517 and jetty-client only being updated with upgrades. I believe that alone might have caused NPE to be thrown.

We actually need only jetty-client now. I've removed jetty-server dependency, upgraded jetty-client to freshly released 9.4.0.v20161208 and couldn't reproduce the issue anymore.

Thanks for looking into this and prompt support.
